### PR TITLE
Fix incorrect directory remove in zero copy replication

### PR DESCRIPTION
**Pull request by alesapin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix extremely rare bug which can lead to potential data loss in zero-copy replication.

### CODE REVIEW

File: src/Storages/MergeTree/DataPartsExchange.cpp [hunk 1]
```diff
@@ -800,7 +801,14 @@ MergeTreeData::MutableDataPartPtr Fetcher::downloadPartToDisk(
...
             data_part_storage_builder->getFullPath());
-        data_part_storage_builder->removeRecursive();
+
+        /// Even if it's a temporary part it could be downloaded with zero copy replication and this function
+        /// is executed as a callback.
+        ///
+        /// We don't control the amount of refs for temporary parts so we cannot decide can we remove blobs
+        /// or not. So we are not doing it
+        bool keep_shared = disk->supportZeroCopyReplication() && data_settings->allow_remote_fs_zero_copy_replication;
+        data_part_storage_builder->removeSharedRecursive(keep_shared);
```
Comment (tavplubix) [inline src/Storages/MergeTree/DataPartsExchange.cpp:811]: Do we need similar fix in `clearOldTemporaryDirectories`?

Comment (alesapin) [reply src/Storages/MergeTree/DataPartsExchange.cpp:811]: Yes

File: src/Storages/MergeTree/MergeTreeData.cpp [hunk 2]
```diff
@@ -1592,7 +1592,17 @@ size_t MergeTreeData::clearOldTemporaryDirectories(size_t custom_directories_lif
...
+                        /// is executed as a callback.
+                        ///
+                        /// We don't control the amount of refs for temporary parts so we cannot decide can we remove blobs
+                        /// or not. So we are not doing it
+
+                        bool keep_shared = disk->supportZeroCopyReplication() && settings->allow_remote_fs_zero_copy_replication;
+                        if (keep_shared)
+                            LOG_WARNING(log, "Since zero-copy replication is enabled we are not going to remove blobs from shared storage for {}", full_path);
+
+                        disk->removeSharedRecursive(it->path(), keep_shared, {});
```
[the remark this answers is not in the document]
Comment (alesapin) [reply src/Storages/MergeTree/MergeTreeData.cpp:1605]: Indeed

**Review by tavplubix — APPROVED** [review]

**Comment by alesapin** [comment]
Integration -- just timeout on start. 
Stateless -- timeout for random test (strange)
memory -- OOM?

**Pull request merged**